doc/rtd/development/datasource_creation.rst

```diff
@@ -0,0 +1,89 @@
+.. _datasource_creation:
+
+Datasource creation
+*******************
+
+Since there are multiple ways to provide `user data`, `meta data`, and
+`vendor data` (each cloud solution seems to prefer its own way), a datasource
+abstract class was internally created to allow for a single way to access the
+different cloud systems methods, providing this data through the typical usage
+of subclasses. See :file:`cloud-init/sources/__init__.py` as a starting
+point for understanding the datastructures used.
+
+The datasource objects have a few touch points with ``cloud-init``. If you
+are interested in adding a new datasource for your cloud platform you will
+need to take care of the following items:
+
+Identify a mechanism for positive identification of the platform
+================================================================
+
+It is good practice for a cloud platform to positively identify itself to
+the guest. This allows the guest to make educated decisions based on the
+platform on which it is running. On the x86 and arm64 architectures, many
+clouds identify themselves through DMI data. For example, Oracle's public
```

Comment:
It wouldn't hurt to link to something external. Not everyone will know.

That said, I'm not sure which link might be best:
[The standard would be most authoritative](https://www.dmtf.org/sites/default/files/standards/documents/DSP0005.pdf).
[Wikipedia might be most accessible](https://en.wikipedia.org/wiki/Desktop_Management_Interface)
[dmidecode may be most familiar to some](https://linux.die.net/man/8/dmidecode)

Reply:
I think probably the source is the best choice. If it were a public-facing doc, rather than a devdoc, I might be tempted to go with a more accessible option, but I think in this case it's better to link to the standard. It's the most likely to be kept up to date.